KIALI_2229 UI - Move lists to virtual scrolls

aljesusg opened the pull request:
[KIALI-2229](https://issues.jboss.org/browse/KIALI-2229)

- [X] Removed pagination
- [X] Removed ItemDescription objects to a 1 file generator
- [X] Move to VirtualScroll
- [X] Move App list
- [X] Move Workloads list
- [X] Move Services list
- [X] Move Istio list

![Screenshot from 2019-07-25 08-36-09](https://user-images.githubusercontent.com/3019213/61851463-5cc25a80-aeb7-11e9-90e4-a9b89a6c4c84.png)

aljesusg commented:
> After installing the PR, I tried to access kiali in browser, but I could not get the login page on hitting the kiali URL. There is no error in kiali pod logs.
> The browser web console shows below error:
> 
> Invariant Violation: "Minified React error #130; visit [https://reactjs.org/docs/error-decoder.html?invariant=130&args[]=undefined&args[]=](https://reactjs.org/docs/error-decoder.html?invariant=130&args%5B%5D=undefined&args%5B%5D=) for the full message or use the non-minified dev environment for full errors and additional helpful warnings. "
> React 16
> e AuthenticationController.tsx:135
> u runtime.js:45
> _invoke runtime.js:264
> e runtime.js:98
> r asyncToGenerator.js:3
> i asyncToGenerator.js:25
> react-dom.production.min.js:3833

@skondkar I think that this is not related with my PR, the error here is in the authentication, which value do you have in the configmap?

hhovsepy commented:
@aljesusg it's set as login, and login with admin/admin passes in fact.

aljesusg commented:
ok thanks @hhovsepy let me check what's going on

hhovsepy commented:
@aljesusg it seems master is broken

jotak commented on src/components/FilterList/FilterComponent.tsx:
```diff
@@ -46,37 +43,6 @@ export abstract class Component<P extends Props<R>, S extends State<R>, R> exten
-    });
-    HistoryManager.setParams([
-      { name: URLParam.PAGE, value: '1' },
-      { name: URLParam.PER_PAGE, value: String(perPage) }
```
These params that are not used anymore can be removed from the enum (in History.ts)

jotak commented on src/components/VirtualList/Config.ts:
```diff
@@ -0,0 +1,109 @@
+  servicerolebinding: { name: 'ServiceRoleBinding', icon: LockedIcon }
+};
+
+export type resource = {
```
Can you use CamelCase for types ? (`Resource`, `ResourceType`, ...)

jotak commented on src/components/VirtualList/VirtualList.tsx:
```diff
@@ -0,0 +1,76 @@
+  type: string;
+};
+
+const listRegex = /\/([a-z0-9-]+)(\/([a-z0-9-]+))?(\/([a-z0-9-]+))?/;
```
Can you add a comment to tell what it's expected to match? I can see it should match something like "/istio" or "/workloads" but i'm not sure if there's more expected.

jotak replied:
also I'm not sure to understand what all these parenthesis are for, except the first ones.

aljesusg replied:
Yeah sorry I was thinking about to control the sort and options inside the table for sort by header feature of PF4 but I'll do later we need the UX design first

jotak commented on src/components/VirtualList/VirtualItem.tsx:
```diff
@@ -0,0 +1,202 @@
+    const { item, index, config, style, className } = this.props;
+    const ns = this.props.item.namespace;
+    return (
+      <Link
```
this link triggers a warning in the console : `Warning: validateDOMNesting(...): <a> cannot appear as a child of <tbody>`
... `in Link (at VirtualItem.tsx:177)`

aljesusg replied:
I know I need to check how to deal with this...virtualtable is not working correctly with listview item.

jotak commented on src/components/VirtualList/Config.ts:
```diff
@@ -0,0 +1,109 @@
+  GlobeRouteIcon,
+  IntegrationIcon,
+  LockedIcon,
+  MigrationIcon,
```
I'm not sure what's wrong with this icon (might be in patternfly) but it triggers a warning in console: `Warning: Invalid DOM property 'class'. Did you mean 'className'?
    in svg (created by MigrationIcon)`

jotak commented on src/components/VirtualList/Config.ts:
```diff
@@ -0,0 +1,109 @@
+  information: resourceType[];
+  caption?: string;
+  istioTypes?: any;
+  icon?: any;
```
To avoid `any`, this can be typed as `IconType` (imported from '@patternfly/react-icons/dist/js/createIcon')

aljesusg replied:
Good one I saw your PR in overview page,  I didn't know about the type

jotak commented on src/components/VirtualList/Config.ts:
```diff
@@ -0,0 +1,109 @@
+  columns?: string[];
+  information: resourceType[];
+  caption?: string;
+  istioTypes?: any;
```
I'm not sure that `istioTypes` should be here in `resource` type, as it's not something that's going to change from a resource to another. What about removing it from resource, exporting the `const istioTypes` and directly consume this exported constant wherever it's used ?

aljesusg replied:
oh you're right

jotak commented on src/components/VirtualList/VirtualList.tsx:
```diff
@@ -0,0 +1,76 @@
+import { config } from './Config';
+
+type VirtualListProps = {
+  rows: any[];
```
I try to see if we can have a better typing here: as it's done with `FilterComponent`, it seems we can use generics here too:
- with here `VirtualListProps<R>` and `rows: R[];`
- on component declaration, with `class VirtualList<R extends TResource>`
- seems like in `render` we should set `tableProps.rows` to empty array instead of `rows`, which is fine as if i understand correctly, the patternfly Table component is not used to display rows but just for the header (is it correct? I tested and it seems true)

skondkar commented:
Verified on:
 Kiali-ui:     1.3.0 (397403e78192766b676f26fa2d435996f9ff4f39)
Kiali:     v1.3.0-SNAPSHOT (5a4f4d00b9b70c5671709f034c35f2cdf203ff40)
Verified scrolling is available and working on all list pages, Verified that pagination is removed. Tested filtering on all list pages. Scrolling is available when it is needed for filter results. Looks good to me.

aljesusg commented:
PR view is ![Screenshot from 2019-07-25 08-36-09](https://user-images.githubusercontent.com/3019213/61851463-5cc25a80-aeb7-11e9-90e4-a9b89a6c4c84.png)


this throw errors like 

```js
index.js:1437 Warning: validateDOMNesting(...): <a> cannot appear as a child of <tbody>.
    in a (created by Context.Consumer)
    in Link (at VirtualItem.tsx:177)
    in VirtualItem (at VirtualTable.tsx:53)
    in CellMeasurer (at VirtualTable.tsx:52)
    in tbody (created by VirtualGrid)
    in table (created by VirtualGrid)
    in VirtualGrid (created by VirtualTableBody)
    in VirtualTableBody (at VirtualTable.tsx:64)
    in div (created by AutoSizer)
    in AutoSizer (at VirtualTable.tsx:62)
    in VirtualTable (at VirtualList.tsx:72)
    in div (at VirtualList.tsx:71)
    in div (at VirtualList.tsx:51)
    in VirtualList (at AppListComponent.tsx:138)
    in AppListComponent (created by Connect(AppListComponent))
    in Connect(AppListComponent) (at AppListPage.tsx:13)
    in AppListPage (created by Context.Consumer)
    in Route (at RenderPage.tsx:13)
    in Switch (at SwitchErrorBoundary.tsx:42)
    in SwitchErrorBoundary (at RenderPage.tsx:28)
    in div (at RenderPage.tsx:27)
    in div (at RenderPage.tsx:39)
    in RenderPage (at Navigation.tsx:116)
    in section (created by PageSection)
    in PageSection (at Navigation.tsx:115)
    in main (created by Page)
    in div (created by Page)
    in Page (at Navigation.tsx:113)
    in Navigation (created by Connect(Navigation))
    in Connect(Navigation) (created by Context.Consumer)
    in Route (created by withRouter(Connect(Navigation)))
    in withRouter(Connect(Navigation)) (at App.tsx:107)
    in Router (at App.tsx:106)
    in AuthenticationController (created by Connect(AuthenticationController))
    in Connect(AuthenticationController) (at App.tsx:117)
    in PersistGate (at App.tsx:115)
    in Provider (at App.tsx:114)
    in App (at src/index.tsx:28)
```

because we can't add this kinds of elements in a table so I'm going to move the design to the proposal and we'll move the filters in another PR to PF4.

cc @jotak @xeviknal @abonas 

Design: https://github.com/kiali/kiali-design/issues/117

aljesusg commented:
![Screenshot from 2019-07-25 13-25-15](https://user-images.githubusercontent.com/3019213/61871040-0f0e1800-aee0-11e9-9803-7fa7a1094844.png)
![Screenshot from 2019-07-25 13-25-04](https://user-images.githubusercontent.com/3019213/61871041-0f0e1800-aee0-11e9-86b9-2add8b0e9007.png)

aljesusg commented:
Opened issue in PF4 https://github.com/patternfly/patternfly-react/issues/2594

cc @serenamarie125 @abonas

jotak commented:
@aljesusg , looking at the examples https://patternfly-react.surge.sh/patternfly-4/virtual%20scroll/virtualized/ , they're always embedding rows/cells in tr/td. Isn't what we should do, too (and keep the Link embedded in cell to handle click) ? Or does it create other issues?

aljesusg commented:
I check how is working the react-virtualized, we can set onclick in the row too. The point is, Do we want set onclick in the row? or Do we want set onclick in the name of application to go to the details, namespace name to go to filter by that ns for example....etc ?

jotak commented:
@aljesusg i've tried this and it seems to work fine:

In VirtualItem
```
    return (
      <tr
        key={config.name + '_ItemItem_' + index + '_' + ns + '_' + item.name}
        style={style}
        className={className}
        role="row"
      >
        <td style={{ padding: 0 }}>
          <Link
            to={this.state.link}
            style={{ color: PfColors.Black }}
          >
// etc.
```
So the click is still handled by the `Link` and we keep control if we want to define other links on namespaces or whatever. What do you think? Maybe I'm missing some problem with that?

aljesusg commented:
@jotak I know that, I mean that now we have a Link in the row, and this is by cell.
I'll change it like OS4 to have the link by cell only in the name of the item instead the row.

aljesusg commented:
Hey @serenamarie125 @abonas What color can we use to let the user know that it is a link? 
Now is PFColors.blue300
![Screenshot from 2019-07-26 12-58-47](https://user-images.githubusercontent.com/3019213/61947303-27943600-afa5-11e9-9979-de86c24e4cfb.png)

abonas commented:
from the UX meeting, the colors to be used are here under the Links section
https://www.patternfly.org/v4/design-guidelines/styles/colors

serenamarie125 commented:
For PF4, this info is on this PF4 page https://www.patternfly.org/v4/design-guidelines/styles/colorshttps://www.patternfly.org/v4/design-guidelines/styles/colors

The global CSS variables are:
--pf-global--link--Color (#0066CC)
--pf-global--link--Color--hover (#004080)

aljesusg commented:
Thanks @serenamarie125 @abonas .

Moved the sort to the table, I did a change in the sort for label validation in workloads.
Before changes we could sort by applabel or versionlabel, in the new design there is a column for both labels so now the sort order the items by:

1. both labels present
2. App present, no version present
3. Version present, no app present
4. no Version, no app so sort by name

Added in services sort by configuration

![Peek 2019-07-30 13-05](https://user-images.githubusercontent.com/3019213/62124558-d1880100-b2ca-11e9-9bb3-46d69d8dc74d.gif)

skondkar commented:
Verified on:
Kiali-ui :    1.3.0 (f6af6a52bca937fa7e8972ce2d90c9ec73e39e5c)
Kiali:     v1.3.0-SNAPSHOT (5e435bc2dc45d90c15b2f336ad315d57826dfa25)
Looks good to me

cfcosta commented on src/components/ConfigValidation/ConfigIndicator.tsx:
```diff
@@ -11,34 +14,38 @@ interface Props {
   color: string;
-  icon: string;
+  icon: IconType;
+  class: string;
```
Just realized this is not a reserved word in JS.

cfcosta commented on src/components/ConfigValidation/ConfigIndicator.tsx:
```diff
@@ -11,34 +14,38 @@ interface Props {
   size?: string;
 }
 
-interface Validation {
```
Since we're exporting this now, can we name it something else, like `ValidationResult`?

cfcosta commented on src/components/ConfigValidation/ConfigIndicator.tsx:
```diff
@@ -151,15 +158,7 @@ export class ConfigIndicator extends React.PureComponent<Props, {}> {
-              tabIndex="0"
-            />
-          </span>
+          <span style={{ color: this.getValid().color }}>{createIcon(this.getValid(), 'sm')}</span>
```
Same for the previous command, this should maybe be named `getValidationResult` or something like that.

cfcosta commented on src/components/Health/Helper.ts:
```diff
@@ -1,8 +1,9 @@
 type Size = 'sm' | 'md' | 'lg' | 'xl';
 
-export const createIcon = (status: Status, size?: Size) => {
+export const createIcon = (status: Status | Validation, size?: Size) => {
```
This seems like a good oportunity to extract those properties in a Interface. Like `InterfaceIcon` or something like that, given that both `Status` and `Validation` have many equal properties.

aljesusg replied:
I was talking with joel about configindicator and health indicator they are so similar and we can extract both...but this is not the target of this PR and we have 900+ now maybe we should open a JIRA for this.

jotak replied:
+1 for another JIRA, this is already quite big. @aljesusg can you create one?

aljesusg replied:
Roger

aljesusg replied:
https://issues.jboss.org/browse/KIALI-3179

cfcosta commented on src/components/VirtualList/VirtualItem.tsx:
```diff
@@ -0,0 +1,75 @@
+  componentWillUnmount() {
+    this.promises.cancelAll();
+  }
+  onItemChanged(item: TResource) {
```
Is this new code or did you move it from somewhere else? If so, can we use `async/await` here? Seems bad adding new code with old-style promises.

aljesusg replied:
This was moved from ItemDescription of workloads, application, istio and services, it's not a new code, I have not problem about move to the new styles, is there any development guide in readme or example about how you want to deal with promises?

cfcosta replied:
I think you can check the examples on `Login.ts` to see what I mean :)

aljesusg commented:
Opened issue about lodash-es in react-table https://github.com/patternfly/patternfly-react/issues/2616

aljesusg commented:
ok PR https://github.com/patternfly/patternfly-react/pull/2641 fix the problem with lodash-es and the PR was updated to this version, please @skondkar can you verify this again ?

@jotak can you review this ? The export of the interfaces will be done in another JIRA

jotak commented:
@aljesusg when several namespaces are selected and we first enter into a view (e.g. services list), it looks like items are sorted by namespace - but the UI shows they should be sorted by name (the name column has the sort icon highlighted)

jotak commented:
Also, it seems sorting is not preserved anymore when switching to another page as before - and not reloaded from URL anymore when we come from a direct link or we refresh the page

jotak commented on src/components/VirtualList/VirtualItem.tsx:
```diff
@@ -0,0 +1,76 @@
+
+  renderDetails = (item: TResource, health?: any) => {
+    const icon = this.getIcon();
+    return this.props.config.columns.map(object => Renderer[object.name](item, this.props.config, icon, health));
```
Calling `Renderer[object.name]` is not ideal here as we lose type safety and compiler enforcement on available renderers. But I understand that changing this would lead to more refactoring, so if you want we can look at that later.

aljesusg replied:
> @aljesusg when several namespaces are selected and we first enter into a view (e.g. services list), it looks like items are sorted by namespace - but the UI shows they should be sorted by name (the name column has the sort icon highlighted)

Oh yes sorry I know where is the problem, I thought that i upload the file

aljesusg replied:
> Calling `Renderer[object.name]` is not ideal here as we lose type safety and compiler enforcement on available renderers. But I understand that changing this would lead to more refactoring, so if you want we can look at that later.

About that it was to be cleaner and simple, only need to add the config and the render method.

jotak commented on src/components/VirtualList/VirtualItem.tsx:
```diff
@@ -0,0 +1,76 @@
+      });
+  };
+
+  renderDetails = (item: TResource, health?: any) => {
```
I believe health could be of type Health , but same as comment below, we can look at that later

aljesusg replied:
yes because typecheck throw nme a warning here

aljesusg commented:
> Also, it seems sorting is not preserved anymore when switching to another page as before - and not reloaded from URL anymore when we come from a direct link or we refresh the page

Yes, I was thinking about....I don't know if it is a good idea to preserve this with this kind of sort, anyway I can add it. What do you think?

jotak commented on src/components/VirtualList/_VirtualList.scss:
```diff
@@ -0,0 +1,70 @@
+.pf-c-scrollablegrid .pf-c-table {
```
Out of curiosity, why is the file prefixed with "_"? I see some file with the same, but most of them without.

aljesusg replied:
I saw it in a guide about good practices and I think that I am the only one who use styles by component, in fact the openshift team use _name too

aljesusg replied:
In that case I imported this file in the index card under styles folder

jotak commented:
>> Also, it seems sorting is not preserved anymore when switching to another page as before - and not reloaded from URL anymore when we come from a direct link or we refresh the page

> Yes, I was thinking about....I don't know if it is a good idea to preserve this with this kind of sort, anyway I can add it. What do you think?

Previously we wanted this (it was a feature), so I guess if we want to remove it, it should be discussed? I don't see new lists as a fundamental change for this feature. @lucasponce , what do you think? Anyway if we don't want the feature anymore, then the related params should be removed from the URL.

aljesusg commented:
> > > Also, it seems sorting is not preserved anymore when switching to another page as before - and not reloaded from URL anymore when we come from a direct link or we refresh the page
> 
> > Yes, I was thinking about....I don't know if it is a good idea to preserve this with this kind of sort, anyway I can add it. What do you think?
> 
> Previously we wanted this (it was a feature), so I guess if we want to remove it, it should be discussed? I don't see new lists as a fundamental change for this feature. @lucasponce , what do you think? Anyway if we don't want the feature anymore, then the related params should be removed from the URL.

I'll add them so

aljesusg commented:
> > > Also, it seems sorting is not preserved anymore when switching to another page as before - and not reloaded from URL anymore when we come from a direct link or we refresh the page
> 
> > Yes, I was thinking about....I don't know if it is a good idea to preserve this with this kind of sort, anyway I can add it. What do you think?
> 
> Previously we wanted this (it was a feature), so I guess if we want to remove it, it should be discussed? I don't see new lists as a fundamental change for this feature. @lucasponce , what do you think? Anyway if we don't want the feature anymore, then the related params should be removed from the URL.

@jotak now If you change from applist to servicelist the sort is not preserved, I enabled the headers when the user set this by queryparams, I was sorting the items but I didn't enable the headers.

@lucasponce I added tooltip for icons

aljesusg commented:
Hooray !!! I'll continue the work in next PRs thanks @jotak @skondkar

jshaughn commented:
@aljesusg Congrats on getting this big job merged!

aljesusg commented:
Thanks @jshauhn it takes so time with the problems with PF4 but finally...:)
